pkg/controller/sparkapplication/controller.go

```diff
@@ -258,91 +259,88 @@ func (c *Controller) processNextItem() bool {
-	nodeName           string         // Name of the node the driver pod runs on.
-	podPhase           apiv1.PodPhase // Driver pod phase.
-	completionTime     metav1.Time    // Time the driver completes.
-	err                error          // Error if any found.
-
-}
-
 func (c *Controller) getAppPods(app *v1beta1.SparkApplication) ([]*apiv1.Pod, error) {
 	// Fetch all the pods for the current run of the application.
 	selector := labels.SelectorFromSet(labels.Set(getResourceLabels(app)))
 	pods, err := c.podLister.Pods(app.Namespace).List(selector)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get pods for SparkApplication %s/%s: %v", app.Namespace, app.Name, err)
 	}
+
 	return pods, nil
 }
 
+// Fetch driver pod from APIServer.
+func (c *Controller) getDriverPodFromApiServer(app *v1beta1.SparkApplication) (*apiv1.Pod, error) {
+	selector := labels.SelectorFromSet(map[string]string{config.SparkAppNameLabel: app.Name, config.SparkRoleLabel: config.SparkDriverRole})
+	listOptions := metav1.ListOptions{LabelSelector: selector.String()}
+	podsList, err := c.kubeClient.CoreV1().Pods(app.Namespace).List(listOptions)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get driver pod for SparkApplication from APIServer %s/%s: %v", app.Namespace, app.Name, err)
+	} else if podsList.Size() == 1 {
+		return &podsList.Items[0], nil
+	}
+	return nil, nil
+}
+
 // getAndUpdateDriverState finds the driver pod of the application
 // and updates the driver state based on the current phase of the pod.
 func (c *Controller) getAndUpdateDriverState(app *v1beta1.SparkApplication) error {
 	pods, err := c.getAppPods(app)
 	if err != nil {
 		return err
 	}
-	var currentDriverState *driverState
+	var driverPod *apiv1.Pod
 	for _, pod := range pods {
 		if util.IsDriverPod(pod) {
-			currentDriverState = &driverState{
-				podName:            pod.Name,
-				nodeName:           pod.Spec.NodeName,
-				podPhase:           pod.Status.Phase,
-				sparkApplicationID: getSparkApplicationID(pod),
-			}
-			if pod.Status.Phase == apiv1.PodSucceeded || pod.Status.Phase == apiv1.PodFailed {
-				currentDriverState.completionTime = metav1.Now()
-			}
-			// Fetch container ExitCode/Reason if Pod Failed.
-			if pod.Status.Phase == apiv1.PodFailed {
-				if len(pod.Status.ContainerStatuses) > 0 && pod.Status.ContainerStatuses[0].State.Terminated != nil {
-					terminatedState := pod.Status.ContainerStatuses[0].State.Terminated
-					currentDriverState.err = fmt.Errorf("driver pod failed with ExitCode: %d, Reason: %s", terminatedState.ExitCode, terminatedState.Reason)
-				} else {
-					currentDriverState.err = fmt.Errorf("driver container status missing.")
-				}
-			}
+			driverPod = pod
 			break
 		}
 	}
+	if driverPod == nil {
+		// Driver Pod was not found in Informer cache. Try to fetch it from APIServer.
+		glog.Infof("driver not found for SparkApplication: %s/%s. Trying to fetch driver from APIServer.", app.Namespace, app.Name)
```

Comment:
Suggesting changing the error message to "driver pod not found for SparkApplication %s/%s, trying to fetch it from the API server.".